### DISCUSSION TITLE
Why am I seeing non-deterministic output in resized image output?

### DISCUSSION THREAD

**Discussion by daiplusplus** [ImageMagick/ImageMagick#3047] [Help]
I'm using ImageMagick `7.0.10-43` via [Magick.NET 7.22.2.2](https://github.com/dlemstra/Magick.NET/releases/tag/7.22.2.2) (Q16, x64) and I'm using it to resize some images down and outputting to an in-memory buffer as JPEG.

today I had been improving my unit-tests and integration tests to include the SHA-256 hash of the output files, and I noticed that despite my tests using the exact same source input image I was getting output from ImageMagick with different SHA-256 hashes. I verified that I am not including any non-deterministic data in the output images (e.g. EXIF metadata when the resized image was saved).

I re-ran my test code several times to generate 8 different output images that should ostensibly be identical and I compared them using BeyondCompare and noticed the the same splodges of pixels were slightly different in all of the different images:

Here's an example: notice the splodges running along the right hand-side of the image. My mouse cursor was hovering over the bottom-right rectangular diff region where one image had a pixel that had RGB `26,31,34` but the other image had RGB `26,31,35`. Other images from the set of 8 had different Red or Green values.

![image](https://user-images.githubusercontent.com/1693078/103229462-e74a8d00-48e7-11eb-902d-f32bb74bf2c9.png)

Is this a bug in ImageMagick, or by-design? If it's by-design, how can I have deterministic output?

## Update:

I changed my code to output the resized images as PNG instead of JPEG to see if the issue was earlier in ImageMagick's pipeline and when I compare PNG output I see the differences are confined to a single column of pixels - so this conceivably is why entire JPEG macroblocks were diff'd - in which case something non-deterministic is happening in ImageMagick's resize code:

![image](https://user-images.githubusercontent.com/1693078/103231153-fcc1b600-48eb-11eb-8f2d-0f6b053a2069.png)

**Comment by dlemstra** [comment]
This is most likely an OpenCL issue. Is the issue resolved when you call `OpenCL.IsEnabled = false;`?

**Comment by daiplusplus** [comment]
`OpenCL.IsEnabled` was `true` and I added a line to set `OpenCL.IsEnabled = false;` however the output is still nondeterministic.

**Comment by daiplusplus** [comment]
I've got a minimum reproduction in LinqPad, if you're interested:

```
<Query Kind="Program">
  <NuGetReference>Magick.NET-Q16-AnyCPU</NuGetReference>
  <Namespace>ImageMagick</Namespace>
  <Namespace>System.Data.Common</Namespace>
  <Namespace>System.Data.SqlClient</Namespace>
  <Namespace>System.Globalization</Namespace>
  <Namespace>System.Runtime.Serialization.Formatters</Namespace>
  <Namespace>System.Security.Cryptography</Namespace>
  <Namespace>System.Threading.Tasks</Namespace>
</Query>

void Main()
{
	Byte[] originalImageData = File.ReadAllBytes( @"Kitten3.jpg" );
	Byte[] originalSha256    = ComputeHash( originalImageData );
	Byte[] resizedSha256;
	
	OpenCL.IsEnabled = false;
	
	using( MemoryStream originalImageStream = new MemoryStream( originalImageData, writable: false ) )
	using( MagickImage image = new MagickImage( stream: originalImageStream ) )
	{
		Byte[] resized = ResizeImageToPng( image, 800, 800 );
		resizedSha256 = ComputeHash( resized );
		
	}
	
	String origHashStr = String.Concat( originalSha256.Select( b => b.ToString("X2") ) );
	String resiHashStr = String.Concat( resizedSha256 .Select( b => b.ToString("X2") ) );
	
	origHashStr.Dump( "Original. " + ( origHashStr == "your-input-hash-here" ? "Matches" : "Diff" ) );
	resiHashStr.Dump( "Resized.  " + ( resiHashStr == "your-output-hash-here" ? "Matches" : "Diff" ) );
	
}

private static Byte[] ComputeHash( Byte[] bytes )
{
	using( SHA256 sha256 = SHA256.Create() )
	{
		return sha256.ComputeHash( bytes );
	}
}

private static Byte[] ResizeImageToPng( MagickImage original, Int32 desiredMaxWidth, Int32 desiredMaxHeight )
{
	using( MagickImage imageResized = new MagickImage( original ) )
	{
		MagickGeometry imSize = new MagickGeometry( width: desiredMaxWidth, height: desiredMaxHeight );
		imageResized.Resize( imSize );

		Byte[] resizedBytes = imageResized.ToByteArray( MagickFormat.Png24 ); //imageResized.ToByteArray( MagickFormat.Jpeg );
		return resizedBytes;
	}
}

```

Every time I run it I get different output resized hashes.

**Comment by dlemstra** [comment]
Can you also include the test image that you are using?

**Comment by daiplusplus** [comment]
There is a minor mistake on my part: I forgot about the PNG `tIME` chunk - so some (but not all) of my test outputs are invalid. Pardon my ignorance, but **how do I save a PNG in Magick.net without the `tIME` chunk?**

I can reproduce the issue with other images though, I'm trying to reduce them down to a reliable repro.

**Comment by daiplusplus** [comment]
@dlemstra Okay I had some wires crossed and the differing output hashes I was seeing with the PNG files was due to the `tIME` chunk when I had `OpenCL.IsEnabled == false`, whereas the differing output hashes I was still getting with JPEG output was when `OpenCL.IsEnabled == true`, but I couldn't reproduce it with JPEG output and `OpenCL.IsEnabled == false`.

So there **definitely is something** in OpenCL, on my machine at least, that's causing nondeterminism.

I can reliably reproduce it with a personal photo of mine, but I can't easily reproduce it with stock photos/images.

Can I email the photo to you directly?

**Comment by dlemstra** [comment]
If it is an OpenCL issue it is most likely something local on your machine (driver or card issue) but maybe we are lucky. You can find my email on my profile page

**Comment by daiplusplus** [comment]
Sent! Thank you!

**Comment by snibgo** [comment]
> ...how do I save a PNG in Magick.net without the tIME chunk?
```
magick in.png -define png:exclude-chunk=tIME out.png
```

See https://imagemagick.org/script/defines.php

**Reply by daiplusplus** [reply]
That's for the `magick` command-line interface, your answer is not helpful at all w.r.t. my question regarding the `Magick.NET` .NET library.

**Reply by dlemstra** [reply]
This can be done with `image.Settings.SetDefine(`

**Comment by snibgo** [comment]
I don't use Magick.NET, and I don't know the equivalent of "-define". Sorry.
